src/dali/plugin/pair_converter/ccxt_converter.py

```diff
@@ -0,0 +1,233 @@
+        super().__init__(historical_price_type=historical_price_type)
+        self.__logger: logging.Logger = create_logger(f"CCXT converter using {historical_price_type}")
+        self.pricing_alt: Dict[str, str] = {}
+        if pricing_alt is not None:
+            self.pricing_alt = json.loads(pricing_alt)
+
+        self.exchanges: Dict[str, Any] = {}
+        self.exchange_markets: Dict[str, List[str]] = {}
+        self.exchange_quote_assets: Dict[str, Set[str]] = {}
+
+    def name(self) -> str:
+        return "CCXT-converter"
+
+    def cache_key(self) -> str:
+        return self.name()
+
+    def get_historic_bar_from_native_source(self, timestamp: datetime, from_asset: str, to_asset: str, exchange: str) -> Optional[HistoricalBar]:
+
+        # If both assets are fiat, skip further processing
+        if from_asset in self.fiat_list and to_asset in self.fiat_list:
+            return self.get_fiat_exchange_rate(timestamp, from_asset, to_asset)
+
+        # Caching of exchanges
+        if exchange not in self.exchanges:
+            if exchange in _EXCHANGEDICT:
+                current_exchange: Any = _EXCHANGEDICT[exchange]()
+                current_markets: List[str] = []
+
+                for market in current_exchange.fetch_markets():
+                    self.__logger.debug("Market: %s", market)
+                    current_markets.append(market[_ID])
+                self.exchanges[exchange] = current_exchange
+                self.exchange_markets[exchange] = current_markets
+            else:
+                self.__logger.error("WARNING: Unrecognized Exchange: %s. Please open an issue at %s", exchange, self.ISSUES_URL)
+                return None
+        else:
+            current_exchange = self.exchanges[exchange]
+            current_markets = self.exchange_markets[exchange]
+
+        ms_timestamp: int = int(timestamp.timestamp() * _MS_IN_SECOND)
+        historical_data: List[List[Union[float, int]]] = []
+        bridge_asset: Optional[str] = None
+        bridge_fiat: Optional[str] = None
+
+        market_symbol = from_asset + to_asset
+
+        # Use the default stable coin if the fiat market doesn't exist
+        # If stable coin market doesn't exist find alt trading pair
+        if market_symbol not in current_markets:
+            if _DEFAULTSTABLEDICT[exchange].fiat_code == to_asset:
+                to_asset = _DEFAULTSTABLEDICT[exchange].symbol
+                market_symbol = from_asset + to_asset
+            elif to_asset in self.fiat_list:
+                bridge_fiat = _DEFAULTSTABLEDICT[exchange].symbol
+                market_symbol = from_asset + bridge_fiat
+
+        result: Optional[HistoricalBar] = None
+        retry_count: int = 0
+
+        # If the asset is not priced in the exchange's default stable coin, alternative pricing must be used.
+        if market_symbol not in current_markets:
+
+            if from_asset in self.pricing_alt:

```

Comment:
I'm following this logic to some extent, but I think this use case is so nuanced that it might require a bit of a write up, if you're up for it. Basically I'm thinking of a page that documents the foreign exchange use case fully for the users. We can add it to the documentation as appropriate (point to it from the FAQ and/or the main docs).

Reply:
Yeah, we need to spell this out since it is very complicated. I need to finalize how exactly the whole mechanism is going to work first. It might even be better to map this out in a diagram.